### PR TITLE
Added item_lmb_selected signal to ItemList

### LINKED ISSUE #23465: ItemList.item_rmb_selected and item_selected signals fire at the same time
Godot 3.1 alpha2,
Godot 3.1 alpha1

ItemList has two signals for item selection:
- `item_selected`
- `item_rmb_selected`

I want to be able to pick up left and right mouse button selections separately to execute a different action. However, when I select with RMB, both signals get sent, which makes me unable to use that second signal... (and by extent, the whole control)

### PR DESCRIPTION
**Pull request by vitika9**
Fix #23465

### CODE REVIEW

**Comment by YeldhamDev** [comment]
Run `--doctool` to update the order in the docs, and then [squash your commits](https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase).

**Comment by KoBeWi** [comment]
Seems like after these changes, rmb selected fires only when clicking on an already selected item (previously it would emit on the first click too) and the new lmb signal has the same problem.
Also if `allow_reselect` is enabled, the lmb signal doesn't emit at all.

**Comment by Vitika9** [comment]
@KoBeWi I have made the required changes.

**Comment by KoBeWi** [comment]
All seems fine now.
But. With `allow_reselect` disabled, `lmb_selected` and `rmb_selected` can still be emitted. It was like that before, so I guess this is expected. But then, I think these signals could be renamed to `lmb_clicked` and `rmb_clicked`, as emitting them without `allow_reselect` can be a bit confusing. Although we should get more opinions on that change probably (even if it makes sense).

File: scene/gui/item_list.cpp [hunk 1]
```diff
@@ -610,26 +610,24 @@ void ItemList::gui_input(const Ref<InputEvent> &p_event) {
...
-					} else if (/*select_mode==SELECT_SINGLE &&*/ mb->is_double_click()) {
-						emit_signal(SNAME("item_activated"), i);
-					}
+				if (mb->get_button_index() == MouseButton::RIGHT) {
+					emit_signal(SNAME("item_rmb_selected"), i, get_local_mouse_position());
+				} else if (mb->get_button_index() == MouseButton::LEFT) {
+					emit_signal(SNAME("item_lmb_selected"), i, get_local_mouse_position());
+				}
+
+				if (/*select_mode==SELECT_SINGLE &&*/ mb->get_button_index() == MouseButton::LEFT && mb->is_double_click()) {
```
Comment (KoBeWi) [inline scene/gui/item_list.cpp:629]: The commented-out code can be removed.

**Comment by Vitika9** [comment]
@KoBeWi OK then, I will keep eyes on tracker

**Review by KoBeWi — APPROVED** [review]
Looks good.

I mentioned renaming the signal, but another thing we could consider is changing the behavior completely xd It's weird to have signal emitted depending on which button was pressed. This could be a signal argument or something.

But well, this is for discussion. The PR also provides some nice refactor that will be useful if we decide to change things further, so it's ok for now.

**Comment by akien-mga** [comment]
I find the current set of signals (both before and after this PR) pretty weird, I also think this might be worth changing. This could be left for a different PR but since this one adds a new signal which makes IMO the problem even more explicit, the discussion is probably worth having here.

The current signals seem to be:
- `item_selected(item)`
- `item_lmb_selected(item, position)` (added by this PR)
- `item_rmb_selected(item, position)`
- `rmb_clicked(position)`

Wouldn't it make more sense to have something like this?
- `item_selected(item)`
- `lmb_clicked(item, position)` (or `left_clicked`) with `item` optionally being `-1` if no item was clicked (i.e. click on background)
- `rmb_clicked(item, position)` (or `right_clicked`), same logic

I haven't reviewed and used the current API in depth so those are just general thoughts on the API.

**Comment by KoBeWi** [comment]
I'd go further and simplify it to:
- `item_selected(item)`
- `item_clicked(item, position, mouse_button)`

**Comment by Vitika9** [comment]
> > I'd go further and simplify it to:
> > 
> > * `item_selected(item)`
> > * `item_clicked(item, position, mouse_button)`
> 
> Makes sense, then maybe there should be an
> 
> * `empty_clicked(position, mouse_button)` or `background_clicked`
> 
> to avoid the confusion with `item_` prefix? For when you click on the background.
> 
> If @Vitika9 wants to try implementing this API, I'd suggest doing it as a separate PR maybe so we can compare both (the scope of changes for this new proposal is likely somewhat bigger). Maybe we should get the design confirmed by other contributors first though.

I would like to try implementing this. Do I have to initiate the conversation in contributor's chat for the design confirmation?

**Comment by akien-mga** [comment]
> I would like to try implementing this. Do I have to initiate the conversation in contributor's chat for the design confirmation?

Yes that's a good idea, you can start a discussion in the #gui channel.

**Comment by Vitika9** [comment]
@akien-mga @KoBeWi 
The problem is that we have two methods in editor_file_dialog.cpp:
* `_item_list_item_rmb_selected` (previously connected to : `item_rmb_selected`, now connected to : `item_selected`) * * * 
* `_item_list_rmb_clicked` (previously connected to : `rmb_clicked`, now connected to : `item_clicked`) 
For second , we can check `mouse_button` (it's third argument) and don't proceed if it isn't right mouse button, but for the first one we don't have any way to check mouse button and we also need `position` in this method.
Shall I change `item_selected(item)` to `item_selected(item, position, mouse_button)`?

**Comment by akien-mga** [comment]
I believe `_item_list_item_rmb_selected` should be connected to `item_clicked`. The click does a selection but that's implicit here, we're mostly interested about the position. `item_selected` is only necessary for when one just wants to know that a selection happens, but not how it happened.

And `_item_list_rmb_clicked` should be connected to `empty_clicked` - that's what `rmb_clicked` was if I'm not mistaken, it fires only when rmb is clicked in empty space (otherwise that's `item_rmb_selected`).

**Comment by Vitika9** [comment]
> I believe `_item_list_item_rmb_selected` should be connected to `item_clicked`. The click does a selection but that's implicit here, we're mostly interested about the position. `item_selected` is only necessary for when one just wants to know that a selection happens, but not how it happened.
> 
> And `_item_list_rmb_clicked` should be connected to `empty_clicked` - that's what `rmb_clicked` was if I'm not mistaken, it fires only when rmb is clicked in empty space (otherwise that's `item_rmb_selected`).

Shall they be renamed to `_item_list_item_rmb_clicked` and `_item_list_rmb_empty_clicked`?

**Comment by akien-mga** [comment]
Superseded by #59720.

**Pull request closed without merging**